app/lib/search/updater.dart

```diff
@@ -4,92 +4,100 @@
 
 import 'dart:async';
 
-import 'package:gcloud/db.dart';
 import 'package:logging/logging.dart';
 
-import '../frontend/models.dart';
-import '../shared/mock_scores.dart';
 import '../shared/search_service.dart';
-import '../shared/utils.dart' show sliceList;
+import '../shared/task_scheduler.dart';
 
-Logger _logger = new Logger('pub.search.updater');
+import 'backend.dart';
+import 'index_ducene.dart';
 
-class PackageIndexUpdater {
-  final PackageIndex index;
-  Timer _timer;
-  bool _updating = false;
+Logger _logger = new Logger('pub.search.updater');
 
-  PackageIndexUpdater(this.index);
+class IndexUpdateTaskSource implements TaskSource {
+  final BatchIndexUpdater _batchIndexUpdater;
+  DateTime _lastTs;
+  IndexUpdateTaskSource(this._batchIndexUpdater);
 
-  Future update({int limit}) async {
-    if (_updating) {
-      _logger.warning('Update called before previous round completed.');
-      return;
-    }
-    try {
-      _updating = true;
-      await _doUpdate(limit);
-    } catch (e, st) {
-      _logger.severe('Error while updating search index.', e, st);
-    } finally {
-      _updating = false;
+  @override
+  Stream<Task> startStreaming() async* {
+    for (;;) {
+      final DateTime now = new DateTime.now().toUtc();
+      int count = 0;
+      await for (String package
+          in searchBackend.listPackages(updatedAfter: _lastTs)) {
+        count++;
+        yield new Task(package, null);
+      }
+      _batchIndexUpdater.reportScanCount(count);
```

Comment:
Ideally we would not build up a new index on each deployment of a new service. We would always have an already built index on cloud storage, which a newly created VM can just load. It can add more newly uploaded packages to that index.

It will periodically write out the newest index to cloud storage, so when the VM gets killed or a new instance gets deployed it can just load it from there.

Also the 'search' service is not making the decision when to migrate traffic. We manually select when it's time to change the default version of a service (for any service, including "search")